src/main/java/pl/damianszczepanik/jenkins/buildhistorymanager/model/actions/DeleteArtifactsWithPatternAction.java

```diff
@@ -0,0 +1,175 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+
+import org.jenkinsci.remoting.RoleChecker;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+import hudson.FilePath;
+import hudson.FilePath.FileCallable;
+import hudson.Util;
+import hudson.model.Run;
+import hudson.remoting.VirtualChannel;
+import jenkins.util.VirtualFile;
+
+/**
+ * Deletes the artifacts with patterns.
+ */
+public class DeleteArtifactsWithPatternAction extends Action {
+    private static final Logger LOG = Logger.getLogger(DeleteArtifactsWithPatternAction.class.getName());
+
+    private String include;
+    private String exclude;
+
+    @DataBoundConstructor
+    public DeleteArtifactsWithPatternAction() {
+        // Jenkins stapler requires to have public constructor with @DataBoundConstructor
+    }
+
+    public String getInclude() {
+        return include;
+    }
+
+    @DataBoundSetter
+    public void setInclude(String include) {
+        this.include = include;
+    }
+
+    public String getExclude() {
+        return exclude;
+    }
+
+    @DataBoundSetter
+    public void setExclude(String exclude) {
+        this.exclude = exclude;
+    }
+
+    static String removeLastSlash(String archiveRootPath) {
+        if(archiveRootPath.endsWith("/") || archiveRootPath.endsWith("\\")) {
+            return archiveRootPath.substring(0, archiveRootPath.length() - 1);
+        } else {
+            return archiveRootPath;
+        }
+    }
+
+    static boolean isDirectoryEmpty(Path path) throws IOException {
+        if (Files.isDirectory(path)) {
+            try (Stream<Path> entries = Files.list(path)) {
+                return !entries.findFirst().isPresent();
+            }
+        }        
+        return false;
+    }
+
+    // if 'file' is on a different node, this FileCallable will be transferred to that node and executed there.
+    public static final class Delete implements FileCallable<Void> {
+        private static final long serialVersionUID = 1L;
+        private final String archiveRootPath;
+
+        public Delete(String archiveRootPath) {
+            this.archiveRootPath = removeLastSlash(archiveRootPath);
+        }
+
+        @Override 
+        public Void invoke(File file, VirtualChannel channel) throws IOException {
+            Set<File> directories = new HashSet<>();
+            deleteFileOrLogError(file, directories);
+            deleteEmptyDirectoriesAndParents(directories);
+            return null;
+        }
+
+        void deleteEmptyDirectoriesAndParents(Set<File> directories) throws IOException {
+            for (File dir : directories) {
+                if (shouldDeleteDirectory(dir)) {
+                    Util.deleteFile(dir);
+                    deleteParentDirectories(dir.getParentFile());
+                }
+            }
+        }
+
+        boolean shouldDeleteDirectory(File dir) throws IOException {
```

Comment:
The default modifier (package visibility) makes it easy to write a direct test of the method from the test class (which lives in the same package).  Otherwise, with `private` we would have a harder time crafting a test that exercises desired code paths in the `shouldDeleteDirectory()` method (or other such helper methods) via one of the other methods visible to the unit tests.  Another alternative is to mark such helper methods `protected` but then we have to have the test class subclass the class under test.  Lastly, there doesn't seem to be much value (yet!) in making this and other helper methods `public`.

Let the record show that a number of these helper methods were extracted due to the code analysis settings flagging methods as "too large" (too many statements and/or too many branches/conditions).

Reply:
Does that answer your question?

Reply:
Or were you asking for a comment to be added?  For example:
```java
        /* default */ boolean shouldDeleteDirectory(File dir) throws IOException {
```

Reply:
Ping!